### PR TITLE
T21756 preparation

### PR DESCRIPTION
**Pull request by mwleeds**
These commits had to broken off from the other T21756 commits because if I add EosUpdaterUtil-0 to `libeos_updater_flatpak_installer_EosUpdaterFlatpakInstaller_@EUFI_API_VERSION@_gir_INCLUDES` in the same commit that adds `EosUpdaterUtil-0.gir` as a target the build fails because the gir isn't installed at build time. So after this is merged, https://github.com/endlessm/eos-updater/pull/188 will be able to have that dependency.

### CODE REVIEW

File: eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8 [hunk 1]
```diff
@@ -10,6 +10,8 @@ eos\-updater\-flatpak\-installer — Endless OS Updater Flatpak Installer
 .IX Header "SYNOPSIS"
 .\"
 \fBeos\-updater\-flatpak\-installer [\-m \fPmode\fB] [\-p]
+.PP
+\fBeos\-updater\-flatpak\-installer \--dry-run
```
Comment (pwithnall) [inline eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: You need to escape each hyphen, so this should end with `\-\-dry\-run`. Isn’t troff/groff fun?

File: eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8 [hunk 2]
```diff
@@ -23,6 +25,11 @@ the old OS deployment is still running.
 .SH OPTIONS
 .IX Header "OPTIONS"
 .\"
+.IP "\fB\-\-dry-run\fP"
```
Comment (pwithnall) [inline eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: Same here.

File: eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8 [hunk 3]
```diff
@@ -23,6 +25,11 @@ the old OS deployment is still running.
...
+Print to stdout the actions that have been applied and the actions left to
```
Comment (pwithnall) [inline eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: No need to mention ‘stdout’. How about “Print the actions that have been applied…”?

File: eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8 [hunk 4]
```diff
@@ -40,7 +47,7 @@ the updated actions to
 When set to \fBcheck\fP, the tool will check to see if all actions are applied (e.g., that
 applications that should have been installed are installed and every app that
 should have been uninstalled is not installed).
-.\"
+\".
```
Comment (pwithnall) [inline eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: Nitpick: This should be in a separate commit.

File: eos-updater-flatpak-installer/main.c [hunk 5]
```diff
@@ -147,6 +149,38 @@ main (int    argc,
...
+                     "Could not get information on which flatpak ref actions to apply: %s",
+                     error->message);
+
+      formatted_flatpak_ref_actions_to_apply =
+        euu_format_all_flatpak_ref_actions ("All flatpak ref actions that are not yet applied",
+                                            new_flatpak_ref_actions_to_apply);
+      g_print ("%s\n", formatted_flatpak_ref_actions_to_apply);
+
+      return EXIT_OK;
+    }
```
[the remark this answers is not in the document]
Comment (pwithnall) [reply eos-updater-flatpak-installer/main.c]: Yeah, duplicating the code paths to implement `--dry-run` isn’t the most maintainable.

File: Makefile.am [hunk 6]
```diff
@@ -283,7 +283,7 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
 libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERSION@_la_LIBADD = \
 	$(CODE_COVERAGE_LIBS) \
 	$(EOS_UPDATER_FLATPAK_INSTALLER_LIBS) \
-	$(top_builddir)/libeos-updater-util/libeos-updater-util-@EUFI_API_VERSION@.la \
+	$(top_builddir)/libeos-updater-util/libeos-updater-util-@EUU_API_VERSION@.la \
```
Comment (pwithnall) [inline Makefile.am]: Good catch. :see_no_evil:

File: Makefile.am [hunk 7]
```diff
@@ -58,7 +59,7 @@ AM_CPPFLAGS = \
 	$(NULL)
 
 # Utility library
-noinst_LTLIBRARIES += libeos-updater-util/libeos-updater-util-@EUU_API_VERSION@.la
+lib_LTLIBRARIES += libeos-updater-util/libeos-updater-util-@EUU_API_VERSION@.la
```
Comment (pwithnall) [inline Makefile.am]: These libraries will also need to be listed in `debian/*.install`, and you will need to add a new subpackage containing the GIR files. There’s some convention for naming the GIR subpackage which I can’t remember, but I’m sure you can cargo-cult it easily enough.

File: Makefile.am [hunk 8]
```diff
@@ -295,6 +296,67 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+	$(libeos_updater_util_libeos_updater_util_@EUU_API_VERSION@_la_SOURCES)
+libeos_updater_util_EosUpdaterUtil_@EUU_API_VERSION@_gir_LIBS = libeos-updater-util/libeos-updater-util-@EUU_API_VERSION@.la
+libeos_updater_util_EosUpdaterUtil_@EUU_API_VERSION@_gir_NAMESPACE = EosUpdaterUtil
+libeos_updater_util_EosUpdaterUtil_@EUU_API_VERSION@_gir_SCANNERFLAGS = \
+	$(WARN_SCANNERFLAGS) \
+	--nsversion=@EUU_API_VERSION@ \
+	--c-include="libeos-updater-util/util.h" \
+	--identifier-prefix=EosUpdater \
+	--identifier-prefix=Euu \
+	--symbol-prefix=euu \
```
Comment (pwithnall) [inline Makefile.am]: Do you also need `--symbol-prefix=eos_updater`?

File: Makefile.am [hunk 9]
```diff
@@ -295,6 +296,67 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+	-I$(top_srcdir) \
```
Comment (pwithnall) [inline Makefile.am]: I’d drop the `-I` here and add `$(libeos_updater_util_libeos_updater_util_@EUU_API_VERSION@_la_CPPFLAGS)` to `libeos_updater_util_EosUpdaterUtil_@EUU_API_VERSION@_gir_CFLAGS`. Ensures they’re built with the same set of preprocessor flags.

File: Makefile.am [hunk 10]
```diff
@@ -295,6 +296,67 @@ libeos_updater_flatpak_installer_libeos_updater_flatpak_installer_@EUFI_API_VERS
...
+	$(NULL)
+
+INTROSPECTION_GIRS += libeos-updater-util/EosUpdaterUtil-@EUU_API_VERSION@.gir
+
+# GObject Introspection for libeos-updater-flatpak-installer
+libeos-updater-flatpak-installer/EosUpdaterFlatpakInstaller-@EUFI_API_VERSION@.gir: libeos-updater-flatpak-installer/libeos-updater-flatpak-installer-@EUFI_API_VERSION@.la
```
Comment (pwithnall) [inline Makefile.am]: Same comments for this GIR file as above.

File: libeos-updater-flatpak-installer/determine-flatpak-actions.c [hunk 11]
```diff
@@ -79,7 +79,18 @@ flatpak_ref_actions_and_progresses (GStrv        directories_to_search,
...
+ * @directories_to_search: (nullable): directories to search for JSON files
+ * @error: return location for a #GError, or %NULL
+ *
+ * Search each directory in @directories_to_search for JSON files (as defined
+ * by eos-updater-flatpak-autoinstall.d(5)) and find actions that should
+ * already have been applied.
+ *
+ * Returns: (element-type filename EuuFlatpakRemoteRefAction) (transfer full):
+ * a mapping from file names to actions that should have been applied
+ */
```
Comment (pwithnall) [inline libeos-updater-flatpak-installer/determine-flatpak-actions.c]: Nitpick: I’d factor all the gtk-doc comment changes out into separate commits, because they’re valuable even if the introspection build changes weren’t to be committed.

File: libeos-updater-flatpak-installer/determine-flatpak-actions.c [hunk 12]
```diff
@@ -97,7 +108,18 @@ eufi_determine_flatpak_ref_actions_to_check (GStrv    directories_to_search,
...
+/**
+ * eufi_determine_flatpak_ref_actions_to_apply:
+ * @directories_to_search: (nullable): directories to search for JSON files
+ * @error: return location for a #GError, or %NULL
+ *
+ * Search each directory in @directories_to_search for JSON files (as defined
+ * by eos-updater-flatpak-autoinstall.d(5)) and find actions that should
+ * be applied.
+ *
+ * Returns: (element-type filename EuuFlatpakRemoteRefAction) (transfer full):
```
[the remark this answers is not in the document]
Comment (pwithnall) [reply libeos-updater-flatpak-installer/determine-flatpak-actions.c]: Nope, `(element-type filename EuuFlatpakRemoteRefAction)` is correct here. [See the documentation](https://wiki.gnome.org/Projects/GObjectIntrospection/Annotations#Type_signature). The `(element-type)` annotation for hash tables has to contain the types of both the key and the value.

Whether `EuuFlatpakRemoteRefAction` is bindable is another question. GIR should be happy with it being listed in the GIR XML as `introspectable="0"`, and the rest of the API will be usable as long as you don’t want to use `eufi_determine_flatpak_ref_actions_to_apply()` from Python.

[the remark this answers is not in the document]
Comment (smspillaz) [reply libeos-updater-flatpak-installer/determine-flatpak-actions.c]: Ah of course. I forgot that this was a hashtable amd has a different syntax. Ignore my comment :)

File: eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8 (hunk 3 above)
Comment (mwleeds) [reply eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: If you just run it on the command line the output is pretty messy because there's other stuff on stderr

Comment (pwithnall) [reply eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: OK.

File: eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8 (hunk 4 above)
Comment (mwleeds) [reply eos-updater-flatpak-installer/docs/eos-updater-flatpak-installer.8]: I don't understand this change or know why it's in there, so I guess I'll take it out.

**Comment by mwleeds** [comment]
I pushed a fixed branch. I would have done fixups but with this small amount of changes I don't think it matters much.

File: debian/control [hunk 13]
```diff
@@ -106,3 +106,18 @@ Description: Updater for Endless OS - utility tests
...
  other update packaging tools.
+
+Package: gir1.2-eos-updater-0
+Section: introspection
+Architecture: any
+Multi-arch: same
+Depends:
+ ${gir:Depends},
+ ${misc:Depends},
+ ${shlibs:Depends},
```
Comment (pwithnall) [inline debian/control]: Not sure the `${shlibs:Depends}` is relevant here, since there aren’t any shared libraries in `gir1.2-eos-updater-0.install`.
